packages/async-loader/async.js

```diff
@@ -1,36 +1,113 @@
-import { h, Component } from 'preact';
+import { h, Component, options, render } from 'preact';
 
-export default function(req) {
-	function Async() {
-		Component.call(this);
+const DOM = '__e';
+const CHILDREN = '__k';
+const PARENT = '__';
+const UNMOUNT = 'unmount';
+const oldUnmountOpts = options[UNMOUNT];
+const oldDiffed = options.diffed;
+const parentNode = document.querySelector('#app');
 
-		let b, old;
-		this.componentWillMount = () => {
-			b = this.base = this.nextBase || this.__b; // short circuits 1st render
-			req(m => {
-				this.setState({ child: m.default || m });
-			});
-		};
+let hydrationNode = null;
+let IS_HYDRATING = false;
+
+const appChildren = [].map.call(
+	parentNode ? parentNode.childNodes : [],
+	elem => elem
+);
+
+const IS_PRERENDERED = appChildren.length > 0;
+
+if (IS_PRERENDERED) {
+	IS_HYDRATING = true;
+
+	options[UNMOUNT] = function(vnode) {
+		/**
+		 *  3. now Pending is being unmount and its dom is being removed
+		 *  so we detach the dom from the vnode and empty its children so that Pending is unmount
+		 *  but the DOM is never actually removed.
+		 */
+		if (vnode.type === Pending) {
+			vnode[DOM] = null;
+			vnode[CHILDREN] = [];
+			// this hook in options is no more needed once hydration is done.
+			options[UNMOUNT] = oldUnmountOpts;
+		}
+		oldUnmountOpts && oldUnmountOpts(vnode);
+	};
+
+	options.diffed = function(vnode) {
+		/**
+		 * We started with an array of all children of #app.
+		 * As preact diffs vnodes top down, we remove their DOM from the array.
+		 * When AsyncComponent is required to render we pick whatever child is on top and render it with dangerouslySetInnerHTML
+		 */
+		if (
+			!hydrationNode &&
+			vnode[DOM] &&
+			vnode[DOM].parentNode === parentNode &&
+			appChildren.includes(vnode[DOM])
+		) {
+			appChildren.shift();
+		}
 
-		this.shouldComponentUpdate = (_, nxt) => {
-			nxt = nxt.child === void 0;
-			if (nxt && old === void 0 && b) {
-				// Node.TEXT_NODE
-				if (b.nodeType === 3) {
-					old = b.data;
-				} else {
-					old = h(b.nodeName, {
-						dangerouslySetInnerHTML: { __html: b.innerHTML },
-					});
-				}
-			} else {
-				old = ''; // dump it
+		/**
+		 *  4. The route component is now contructed and its DOM will be appended to the browser's DOM.
+		 *  But right before it, we swap its newly contructed DOM with the DOM already present on the browser.
+		 *  TLDR; This prevent a duplicate DOM entry
+		 */
+		if (
+			vnode[DOM] &&
+			vnode[DOM].parentNode === null &&
+			vnode[PARENT] &&
+			vnode[PARENT][PARENT] &&
+			vnode[PARENT][PARENT].type.name === 'AsyncComponent'
+		) {
+			vnode[DOM] = hydrationNode;
+			// this hook in options is no more needed once hydration is done.
+			options.diffed = oldDiffed;
+		}
+		oldDiffed && oldDiffed(vnode);
+	};
+}
+
+function Pending() {
+	hydrationNode = appChildren.length > 0 ? appChildren[0] : null;
```

Comment:
Hmm - searching within the app's root node is going to fail in a bunch of cases, like if there are nodes between App and the Router. Maybe we could inject a prop during prerendering and then look for it during hydration?

Reply:
can you give a small example of what this failing case would look like?
e.g. this can handle 
```jsx
<App>
  <header/>
  <nav />
  <Router>
    ...
  </Router>
</App>
```

Reply:
```js
<App>
  <div>
    <header/>
    <nav />
    <Router>
      ...
    </Router>
  </div>
</App>
```

Reply:
Right now during prerendering we bypass `async-loader`, but we could instead inject a component that marks AsyncComponent boundaries :

```js
exports.pitch = function(req) {
	this.cacheable && this.cacheable();
	
	return `
		import Sync from '@preact/async-loader/sync';
		function load(cb) {
			cb( require(${stringifyRequest(this, '!!' + req)}) );
		}
		export default Sync(load);
	`;
};
```

```js
import { h, options } from 'preact';

var next, old = options.vnode;
options.vnode = function(vnode) {
  var x = 'type' in vnode;
  var tag = x ? vnode.type : vnode.nodeName;
  var props = x ? 'props' : 'attributes';
  if (next && typeof tag === 'string') {
    next = null;
    (vnode[props] || (vnode[props] = {}))['data-pacr'] = next;
  }
};

export default function sync(load, name) {
  var component;
  return function SyncComponent(props) {
    if (!component) component = load();
    next = name;
    return h(component, props);
  }
}
```

Then the client could pick up the root using `querySelector("[data-pacr]").localName`.

Ideally this wouldn't be necessary. I wonder if there's an options hook we could use that would fire just prior to rendering the AsyncComponent and include the candidate DOM node?  That was the thing I stumbled on when porting this from 8 to X - in 8, we had `component._nextBase`.

Reply:
I forgot to make the point in my comment, but the reason for the complexity is because technically `async-component`/`async-loader` can be used anywhere in the component tree. In CLI that's anywhere you do `import X from 'async!./x'`.

Reply:
Instead of holding everything back, should we temporarily drop support for `async!` and move ahead for automatic code splitting for `routes`?

Now that almost everyone is more used to Suspense anyways for code splitting?

Reply:
We could drop support for pre-rendering `async!` and switch it to just boot only on the client. That way we're not dropping the split point and causing a huge change in bundle structure, but we're not holding back on progressive hydration.

Reply:
Nit: Following code doesn't work
```js
options.vnode = function(vnode) {
  var x = 'type' in vnode;
  var tag = x ? vnode.type : vnode.nodeName;
  var props = x ? 'props' : 'attributes';
  if (next && typeof tag === 'string') {
    next = null;
    (vnode[props] || (vnode[props] = {}))['data-pacr'] = next;
  }
};
```

because the vnodes are constructed bottom up, so `if (next && typeof tag === 'string') {` does not point to the actual AsyncBoundary  but the farthest child of it.